community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreAbstractNodeCursor.java

```diff
@@ -19,61 +19,502 @@
+                {
+                    RelationshipItem rel = relationship.get();
+
+                    int[] byType = degrees.get( rel.type() );
+                    if ( byType == null )
+                    {
+                        degrees.put( rel.type(), byType = new int[3] );
+                    }
+                    byType[directionOf( nodeRecord.getId(), rel.id(), rel.startNode(), rel.endNode() ).ordinal()]++;
+                }
+            }
+
+            final PrimitiveIntIterator keys = degrees.iterator();
+
+            return new DegreeItemIterator( keys, degrees );
+        }
+    }
+
+    private long nodeDegreeByDirection( RelationshipGroupRecord group, Direction direction )
+    {
+        long loopCount = countByFirstPrevPointer( group.getFirstLoop() );
+        switch ( direction )
+        {
+            case OUTGOING:
+                return countByFirstPrevPointer( group.getFirstOut() ) + loopCount;
+            case INCOMING:
+                return countByFirstPrevPointer( group.getFirstIn() ) + loopCount;
+            case BOTH:
+                return countByFirstPrevPointer( group.getFirstOut() ) +
+                        countByFirstPrevPointer( group.getFirstIn() ) + loopCount;
+            default:
+                throw new IllegalArgumentException( direction.name() );
+        }
+    }
+
+    private long countByFirstPrevPointer( long relationshipId )
+    {
+        if ( relationshipId == Record.NO_NEXT_RELATIONSHIP.intValue() )
+        {
+            return 0;
+        }
+        RelationshipRecord record = neoStore.getRelationshipStore().getRecord( relationshipId );
+        if ( record.getFirstNode() == nodeRecord.getId() )
+        {
+            return record.getFirstPrevRel();
+        }
+        if ( record.getSecondNode() == nodeRecord.getId() )
+        {
+            return record.getSecondPrevRel();
+        }
+        throw new InvalidRecordException( "Node " + nodeRecord.getId() + " neither start nor end node of " + record );
+    }
+
+    private Direction directionOf( long nodeId, long relationshipId, long startNode, long endNode )
+    {
+        if ( startNode == nodeId )
+        {
+            return endNode == nodeId ? Direction.BOTH : Direction.OUTGOING;
+        }
+        if ( endNode == nodeId )
+        {
+            return Direction.INCOMING;
+        }
+        throw new InvalidRecordException( "Node " + nodeId + " neither start nor end node of relationship " +
+                relationshipId + " with startNode:" + startNode + " and endNode:" + endNode );
+    }
+
+    private static class DegreeItemIterator implements Cursor<DegreeItem>, DegreeItem
+    {
+        private final PrimitiveIntObjectMap<int[]> degrees;
+        private PrimitiveIntIterator keys;
+
+        private int type;
+        private int outgoing;
+        private int incoming;
+
+        public DegreeItemIterator( PrimitiveIntIterator keys, PrimitiveIntObjectMap<int[]> degrees )
```

Comment:
Seems to be quite wasteful data structures here since by the looks of usage of this constructor the keys will correlate to the keys in the map anyway, so a simple array of arrays, or even just a flat array with four items per degree (`type`, `out`, `in`, `loop`) should suffice and be much cheaper.